TaffoInitializer/IndirectCallPatcher.cpp

```diff
@@ -0,0 +1,174 @@
+#include "IndirectCallPatcher.h"
+
+#include "Metadata.h"
+#include "TaffoInitializerPass.h"
+#include "TypeUtils.h"
+#include "llvm/ADT/Statistic.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+#include <llvm/IR/IRBuilder.h>
+#include <llvm/Transforms/Utils/Cloning.h>
+#include <llvm/Transforms/Utils/ValueMapper.h>
+
+using namespace taffo;
+using namespace llvm;
+
+/// Check recursively whether an unsupported function is called.
+bool containsUnsupportedFunctions(llvm::Function *function, int exploreDepth) {
+  std::vector<std::string> prefixBlocklist{"__kmpc_omp_task"};
+
+  for (auto instructionIt = inst_begin(function);
+       instructionIt != inst_end(function); instructionIt++) {
+    if (auto curCallInstruction = dyn_cast<CallInst>(&(*instructionIt))) {
+      llvm::Function *curCallFunction = curCallInstruction->getCalledFunction();
+      auto functionName = curCallFunction->getName();
+
+      if (any_of(prefixBlocklist, [&](const std::string &prefix) {
+            return functionName.startswith(prefix);
+          })) {
+        return true;
+      } else if (exploreDepth > 0) {
+        if (containsUnsupportedFunctions(curCallFunction, exploreDepth - 1)) {
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+
+/// Handle the __kmpc_fork_call, by replacing the indirect call with a direct call to the function.
+/// In case an unsupported function is called, be conservative and keep the indirect function.
+void handleKmpcFork(const Module &m, std::vector<Instruction *> &toDelete,
+                    CallInst *curCallInstruction, const CallSite *curCall,
+                    Function *indirectFunction) {
+  auto microTaskOperand =
+      dyn_cast<ConstantExpr>(curCall->arg_begin() + 2)->getOperand(0);
+  auto microTaskFunction = dyn_cast<Function>(microTaskOperand);
```

Comment:
`llvm::dyn_cast<>()` should be preceded by a check with `llvm::isa<>()` to make sure it does not crash.
An alternative solution is to use `llvm::dyn_cast_or_null<>()` which returns `nullptr` in case of failed cast, just like `std::dynamic_cast<>()`.
Anyway, before or after these casts there should be a fail-safe check.

Reply:
@skeru I think you are mistaking `llvm::dyn_cast<>()` with `llvm::static_cast<>()`. The former does not crash if its argument has the wrong type, but returns `nullptr`, just like `std::dynamic_cast<>()`.

On the contrary, you should check whether `microTaskOperand` and `microTaskFunction` are `nullptr` before using them.
If you are sure they are instances of the right class and they are not `nullptr` by some function pre-condition, then use `llvm::static_cast<>()`. If the pre-condition is not met, then crashing is the right thing to do, because the error is somewhere else.

Reply:
`llvm::dyn_cast<>()` does not allow its argument to be `nullptr`, `llvm::dyn_cast_or_null<>()` does.

See https://www.llvm.org/docs/ProgrammersManual.html#the-isa-cast-and-dyn-cast-templates

Reply:
I know, but you wrote "`llvm::dyn_cast<>()` should be preceded by a check with `llvm::isa<>()` to make sure it does not crash."

`llvm::isa<>()` still crashes if the argument is `nullptr`.

Reply:
Used dyn_cast_or_null for the second cast to account for possible casting errors in the first cast.

In this way, I have only a check afterwards to make sure it is not `nullptr`.
In particular, the pre-condition is the signature of the OpenMP runtime library function described in the LLVM reference, therefore I crash with an `assert` in case the output of the dyn_cast is a `nullptr`.

This has the con of not being conservative in case OpenMP changes the library runtime functions, but in case that happens the whole OpenMP integration would need a revamp.

The alternative would be to extremely careful and conservative, but it can go only so far in case of function signature changes since the shared variables would be undetected and therefore give unexpected behaviors.

Reply:
Appears fixed to me :D